lib/beaker/hypervisor/abs.rb

```diff
@@ -22,28 +23,83 @@ def provision
       #   "type"     => "centos-7-i386",
       #   "engine"   => "vmpooler",
       # }
-      @resource_hosts.each do |resource_host|
-        type = resource_host['type']
-        type2hosts[type] ||= []
-        type2hosts[type] << resource_host['hostname']
-      end
+      type2hosts = @resource_hosts.group_by { |resource_host| resource_host['type'] }
+
+      missing_templates = []
+      missing_hosts = []
 
       # for each host, get a vm for that template type
       @hosts.each do |host|
         template = host['template']
 
-        raise ArgumentError.new("Failed to provision host '#{host.hostname}' because its 'template' is missing.") if template.nil?
+        if template.nil?
+          missing_templates << host
+          next
+        end
 
         if provisioned_hosts = type2hosts[template]
-          host['vmhostname'] = provisioned_hosts.shift
+          host['vmhostname'] = provisioned_hosts.shift['hostname']
+          tag(host)
```

Comment:
We should only tag hosts whose `engine` is `vmpooler`, otherwise we might try to tag a hardware resource.